### PR TITLE
Extend cache on database failure

### PR DESCRIPTION
**Pull request by angelim**
Allows consumers to reuse(and extend) an old cache when the database is not available for a refresh. If the cache is not yet bootstrapped we still raise the exception.

This feature is opt-in and managed by `Arturo::Feature.extend_cache_on_failure` config.

Logging(optional) was added to help debug cache extension events.
You'll be able to see one of:

* [Arturo][extend_cache_on_failure] Fallback cache is empty
* [Arturo][extend_cache_on_failure] Falling back to stale cache

**Side note:**

Also had to lock sqlite3 on older rails gemfiles to make travis happy.
```
# Gem::LoadError:
#   can't activate sqlite3 (~> 1.3.5), already activated sqlite3-1.4.2. Make sure all dependencies are added to Gemfile.
#   ./spec/support/prepare_database.rb:4:in `<top (required)>'
```

### CODE REVIEW

File: lib/arturo/feature_caching.rb [hunk 1]
```diff
@@ -60,9 +61,8 @@ def fetch(cache, symbol, &block)
           features = cache.read("arturo.all")
 
           unless cache_is_current?(cache, features)
-            features = Hash[Arturo::Feature.all.map { |f| [f.symbol.to_sym, f] }]
-            mark_as_current!(cache)
-            cache.write("arturo.all", features, :expires_in => Arturo::Feature.cache_ttl * 10)
+            features = arturos_from_origin(features)
```
Comment (craig-day) [inline lib/arturo/feature_caching.rb]: can we rename both of these variables? Maybe the method argument could be `existing_features` and the new variable could be `updated_features`? reusing the name `features` everywhere makes the code a little harder to follow.

File: lib/arturo/feature_caching.rb [hunk 2]
```diff
@@ -74,16 +74,40 @@ def expire(cache, symbol)
 
         private
 
+        def arturos_from_origin(fallback_features)
+          Hash[Arturo::Feature.all.map { |f| [f.symbol.to_sym, f] }]
+        rescue ActiveRecord::ActiveRecordError
+          Arturo::Feature.extend_cache_on_failure ? fallback_features : raise
```
Comment (craig-day) [inline lib/arturo/feature_caching.rb]: nit: I prefer a guard style raise

```rb
rescue ActiveRecord::ActiveRecordError
  raise unless Arturo::Feature.extend_cache_on_failure
  fallback_features
end
```

File: lib/arturo/feature_caching.rb [hunk 3]
```diff
@@ -74,16 +74,40 @@ def expire(cache, symbol)
...
-          return false if features.values.map(&:updated_at).compact.max != Arturo::Feature.maximum(:updated_at)
+          begin
+            return false if origin_changed?(features)  
+          rescue ActiveRecord::ActiveRecordError
+            if Arturo::Feature.extend_cache_on_failure
+              update_and_extend_cache!(cache, features)
+              return true
+            else
+              raise
+            end
```
Comment (craig-day) [inline lib/arturo/feature_caching.rb]: similar to above, maybe something like

```rb
rescue ActiveRecord::ActiveRecordError
  raise unless Arturo::Feature.extend_cache_on_failure
  update_and_extend_cache!(cache, features)
  true
end
```

File: spec/models/cache_spec.rb [hunk 4]
```diff
@@ -196,5 +196,81 @@ def lock_down_maximum
         Arturo::Feature.to_feature(@feature.symbol)
       }.to_not raise_error
     end
+
+    describe 'database errors' do
+      before do
+        Arturo::Feature.to_feature(@feature.symbol)
+        @feature.touch
+        Timecop.travel(Time.now + Arturo::Feature.cache_ttl + 5.seconds)
```
Comment (craig-day) [inline spec/models/cache_spec.rb]: Is time already frozen? I don't see it in the highest level before block.

File: lib/arturo/feature_caching.rb [hunk 5]
```diff
@@ -36,12 +36,13 @@ def to_feature(feature_or_symbol)
...
       class << base
         prepend PrependMethods
-        attr_accessor :cache_ttl, :feature_cache, :feature_caching_strategy
+        attr_accessor :cache_ttl, :feature_cache, :feature_caching_strategy, :extend_cache_on_failure
       end
       base.send(:after_save) do |f|
         f.class.feature_caching_strategy.expire(f.class.feature_cache, f.symbol.to_sym) if f.class.caches_features?
       end
       base.cache_ttl = 0
+      base.extend_cache_on_failure = false
```
Comment (vanchi-zendesk) [inline lib/arturo/feature_caching.rb]: Is it better to have it on by default?

Comment (angelim) [reply lib/arturo/feature_caching.rb]: I didn't want to change the current behavior to avoid any surprises

File: spec/models/cache_spec.rb (hunk 4 above)
Comment (angelim) [reply spec/models/cache_spec.rb]: It wasn't frozen before, but we need it to travel to any point after TTL to force a refresh.

File: lib/arturo/feature_caching.rb [hunk 6]
```diff
@@ -36,12 +36,13 @@ def to_feature(feature_or_symbol)
     def self.extended(base)
       class << base
         prepend PrependMethods
-        attr_accessor :cache_ttl, :feature_cache, :feature_caching_strategy
+        attr_accessor :cache_ttl, :feature_cache, :feature_caching_strategy, :extend_cache_on_failure
```
Comment (bquorning) [inline lib/arturo/feature_caching.rb]: I think I would prefer only a writer method for `:extend_cache_on_failure`, and then add a `#extend_cache_on_failure?` method as the reader, e.g.
```rb
def extend_cache_on_failure?
  !!@extend_cache_on_failure
end
```

Comment (angelim) [reply lib/arturo/feature_caching.rb]: good idea. Done

**Comment by angelim** [comment]
@craig-day need another review. Had to introduce logging here.

**Pull request merged**